lib/internal/errors.js

```diff
@@ -0,0 +1,282 @@
+// include their own alternatives to this because those need to be portable
+// to the readable-stream standalone module. If the logic here changes at all,
+// then those also need to be checked.
+
+const kCode = Symbol('code');
+
+class NodeError extends Error {
+  constructor(key, ...args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeError);
+  }
+
+  get name() {
+    return `Error[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+class NodeTypeError extends TypeError {
+  constructor(key, ...args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeTypeError);
+  }
+
+  get name() {
+    return `TypeError[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+class NodeRangeError extends RangeError {
+  constructor(key, ... args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeRangeError);
+  }
+
+  get name() {
+    return `RangeError[${this[kCode]}]`;
+  }
+
+  get code() {
+    return this[kCode];
+  }
+}
+
+exports.Error = NodeError;
+exports.TypeError = NodeTypeError;
+exports.RangeError = NodeRangeError;
+
+var assert, util;
+function lazyAssert() {
+  if (!assert)
+    assert = require('assert');
+  return assert;
+}
+
+function lazyUtil() {
+  if (!util)
+    util = require('util');
+  return util;
+}
+
+function message(key, args) {
+  const assert = lazyAssert();
+  const util = lazyUtil();
+  const msg = Error[Symbol.for(key)];
+  // We want to assert here because if this happens,
+  // Node.js is doing something incorrect.
+  assert(msg, `An invalid error message key was used: ${key}.`);
+  let fmt = util.format;
+  if (typeof msg === 'function') {
+    fmt = msg;
+  } else {
+    args.unshift(msg);
+  }
+  const res = fmt.apply(null, args);
+  // We want to assert here because if this happens,
+  // Node.js is doing something incorrect.
+  assert(res, 'An invalid formatted error message was returned.');
+  return res;
+}
+
+// Below is the catalog of messages
+// Error keys should be all caps, value can
+// be a string formatted for use with
+// util.format, or a function.
+
+// Utility function for registering the error codes.
+function E(sym, val) {
+  Error[Symbol.for(sym)] = val;
+}
+
+E('ARGNOTNUM', argNotNum);
+E('ASSERTIONERROR', (msg) => msg);
+E('BUFFERLIST',
+  '"buffer" list argument must contain only Buffers or strings');
+E('BUFFERSWAP16', 'Buffer size must be a multiple of 16-bits');
+E('BUFFERSWAP32', 'Buffer size must be a multiple of 32-bits');
+E('BUFFERTOOLARGE', bufferTooBig);
+E('CALLBACKREQUIRED', 'The required "callback" argument is not a function');
+E('CHANNELCLOSED', 'Channel closed');
+E('CLOSEDBEFOREREPLY', 'Slave closed before reply');
+E('CONSOLEWRITABLE', 'Console expects a writable stream instance');
+E('DNSUNKNOWNDOMAIN', 'Unable to determine the domain name');
```

Comment:
I was gonna get creative with 'em but figured they were crazy enough as it was. I'm open to suggestions for a more sensible naming convention.